xds/src/main/java/io/grpc/xds/OrcaOobUtil.java

```diff
@@ -0,0 +1,627 @@
+            overallConfig = overallConfig.toBuilder()
+                .setReportInterval(minInterval, TimeUnit.NANOSECONDS).build();
+            reconfigured = true;
+          }
+        }
+        if (reconfigured) {
+          stopRpc("ORCA reporting reconfigured");
+          adjustOrcaReporting();
+        }
+      }
+
+      @Override
+      public void onSubchannelState(ConnectivityStateInfo newState) {
+        if (Objects.equal(state.getState(), READY) && !Objects.equal(newState.getState(), READY)) {
+          // A connection was lost.  We will reset disabled flag because ORCA service
+          // may be available on the new connection.
+          disabled = false;
+        }
+        if (Objects.equal(newState.getState(), SHUTDOWN)) {
+          orcaHelper.orcaStates.remove(this);
+        }
+        state = newState;
+        adjustOrcaReporting();
+        // Propagate subchannel state update to downstream listeners.
+        stateListener.onSubchannelState(newState);
+      }
+
+      void adjustOrcaReporting() {
+        if (!disabled && overallConfig != null && Objects.equal(state.getState(), READY)) {
+          if (orcaRpc == null && !isRetryTimerPending()) {
+            startRpc();
+          }
+        } else {
+          stopRpc("Client stops ORCA reporting");
+          backoffPolicy = null;
+        }
+      }
+
+      void startRpc() {
+        checkState(orcaRpc == null, "previous orca reporting RPC has not been cleaned up");
+        checkState(subchannel != null, "init() not called");
+        subchannelLogger.log(
+            ChannelLogLevel.DEBUG, "Starting ORCA reporting for {0}", subchannel.getAllAddresses());
+        orcaRpc = new OrcaReportingStream(subchannel.asChannel(), stopwatchSupplier.get());
+        orcaRpc.start();
+      }
+
+      void stopRpc(String msg) {
+        if (orcaRpc != null) {
+          orcaRpc.cancel(msg);
+          orcaRpc = null;
+        }
+        if (retryTimer != null) {
+          retryTimer.cancel();
+          retryTimer = null;
+        }
+      }
+
+      boolean isRetryTimerPending() {
+        return retryTimer != null && retryTimer.isPending();
+      }
+
+      @Override
+      public String toString() {
+        return MoreObjects.toStringHelper(this)
+            .add("disabled", disabled)
+            .add("orcaRpc", orcaRpc)
+            .add("reportingConfig", overallConfig)
+            .add("connectivityState", state)
+            .toString();
+      }
+
+      private class OrcaReportingStream extends ClientCall.Listener<OrcaLoadReport> {
+
+        private final ClientCall<OrcaLoadReportRequest, OrcaLoadReport> call;
+        private final Stopwatch stopwatch;
+        private boolean callHasResponded;
+
+        OrcaReportingStream(Channel channel, Stopwatch stopwatch) {
+          call =
+              checkNotNull(channel, "channel")
+                  .newCall(OpenRcaServiceGrpc.getStreamCoreMetricsMethod(), CallOptions.DEFAULT);
```

Comment:
What's the reason for using the low level APIs `channel.newCall`, `call.start()`, `call.halfClose()`, `call.request(1)` and etc rather than using Stub and `ClientInterceptor` instead?

Reply:
This is a server streaming call. Calls cannot be canceled from client side if using the stub approach.

Reply:
There is a way to cancel the call using stub, you can cast the `requestObserver` returned by the stub to `ClientCallStreamObserver` type. It is widely used for cancellation, per message compression and flow control.